rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -120,26 +107,26 @@ impl PageScheduler for BinaryPageScheduler {
         // if user wants row range a..b
         // Case 1: if a != 0, we need indices a-1..b to decode
         // Case 2: if a = 0, we need indices 0..b to decode
-        let indices_ranges = ranges
+        // To get the byte ranges we then multiply these by 8
+        // Then we add the buffer offset to map to the correct buffer in the page,
+        // since multiple encoding tasks may have been used to create the page
+        let indices_byte_ranges = ranges
             .iter()
             .map(|range| {
                 if range.start != 0 {
-                    (range.start - 1)..(range.end)
+                    (self.buffer_offset + ((range.start - 1) * 8))
+                        ..(self.buffer_offset + (range.end * 8))
                 } else {
-                    0..(range.end)
+                    self.buffer_offset..(self.buffer_offset + (range.end * 8))
                 }
             })
             .collect::<Vec<std::ops::Range<u64>>>();
 
         let num_rows = ranges.iter().map(|r| r.end - r.start).sum::<u64>();
 
-        let mut futures_ordered = indices_ranges
-            .iter()
-            .map(|range| {
-                self.indices_scheduler
-                    .schedule_ranges(&[range.clone()], scheduler, top_level_row)
-            })
-            .collect::<FuturesOrdered<_>>();
+        // We schedule all the indices for decoding together
+        // This is more efficient compared to scheduling them one by one (reduces speed significantly for random access)
+        let indices_bytes = scheduler.submit_request(indices_byte_ranges, top_level_row);
```

Comment:
I think you can combine these into one call and keep `indices_scheduler`.  Just make a single call to `indices_scheduler.schedule_ranges` (passing in many ranges) instead of many calls to `schedule_ranges` (each passing in one range).